File: race_mgmt_python/race/interfaces/vehicle.py
# This file contains the abstract class Vehicle which is the parent class for all cars and bikes
# Author: Cameron Denton
# Created: 1/14/2025
# Last Modified: 1/14/2025

# Import the ABC and abstractmethod from the abc module
from abc import ABC, abstractmethod

# Create the abstract class Vehicle
class Vehicle(ABC):

    # Constructor for the Vehicle class
    def __init__(self, vehicle_type, name, start_speed, max_speed, acceleration_rate, position):
        self._type = vehicle_type
        self._name = name
        self._start_speed = start_speed
        self._max_speed = max_speed
        self._acceleration_rate = acceleration_rate
        self._position = position

    # Abstract method to start the vehicle
    @abstractmethod
    def start(self):
        pass

    # Abstract method to stop the vehicle
    @abstractmethod
    def drive(self):
        pass

    # Returns the Vehicle type
    def Type(self):
        return self._type
    
    # Returns the Vehicle name
    def Name(self):
        return self._name
    
    # Returns the Vehicle start speed
    def Start_Speed(self):
        return self._start_speed
    
    # Returns the Vehicle start speed
    def Max_Speed(self):
        return self._max_speed
    
    # Returns the Vehicle max speed
    def Acceleration_Rate(self):
        return self._acceleration_rate

    # Returns the Vehicle position
    def Position(self):
        return self._position
    
    # Updates the Vehicle position
    def Update_Position(self, position):
        self._position = position
    

File: race_mgmt_go/internal/race/interfaces/vehicle.go
package interfaces

// Vehicle interface defines the required methods that all vehicles must implement.
type Vehicle interface {
	Type() string
	Get_Name() string
	Start_Speed() int
	Max_Speed() int
	Acceleration_Rate() int
	Get_Position() int
	Update_Position(int) int
	Start() int
	Drive() int
}

// BaseVehicle struct provides a base implementation for common fields and methods
// shared by all vehicle types.
type BaseVehicle struct {
	VehicleType string
	Name        string
	StartSpeed  int
	MaxSpeed    int
	AccelRate   int
	Position    int
}

// Type returns the type of the vehicle (e.g., "Car", "Bike").
func (v *BaseVehicle) Type() string {
	return v.VehicleType
}

// Name returns the name of the vehicle (e.g., "Ferrari Enzo").
func (v *BaseVehicle) Get_Name() string {
	return v.Name
}

// Start_Speed returns the initial starting speed of the vehicle.
func (v *BaseVehicle) Start_Speed() int {
	return v.StartSpeed
}

// Max_Speed returns the maximum speed the vehicle can achieve.
func (v *BaseVehicle) Max_Speed() int {
	return v.MaxSpeed
}

// Acceleration_Rate returns the rate at which the vehicle accelerates.
func (v *BaseVehicle) Acceleration_Rate() int {
	return v.AccelRate
}

// Position returns the current position of the vehicle in the race.
func (v *BaseVehicle) Get_Position() int {
	return v.Position
}

// Update_Position updates the vehicle's position by adding the given distance
// and returns the updated position.
func (v *BaseVehicle) Update_Position(distance int) int {
	v.Position += distance
	return v.Position
}
